ecommerce/extensions/payment/views/liqpay.py

```diff
@@ -0,0 +1,169 @@
+from ecommerce.extensions.payment.exceptions import DuplicateReferenceNumber, InvalidBasketError, InvalidSignatureError
+from ecommerce.extensions.payment.processors.liqpay import Liqpay
+
+
+logger = logging.getLogger(__name__)
+
+Applicator = get_class('offer.applicator', 'Applicator')
+Basket = get_model('basket', 'Basket')
+BillingAddress = get_model('order', 'BillingAddress')
+Country = get_model('address', 'Country')
+NoShippingRequired = get_class('shipping.methods', 'NoShippingRequired')
+Order = get_model('order', 'Order')
+OrderNumberGenerator = get_class('order.utils', 'OrderNumberGenerator')
+OrderTotalCalculator = get_class('checkout.calculators', 'OrderTotalCalculator')
+PaymentProcessorResponse = get_model('payment', 'PaymentProcessorResponse')
+
+
+class LiqpayPaymentCallbackView(EdxOrderPlacementMixin, View):
+    """Execute an approved LiqPay payment and place an order for paid products as appropriate."""
+
+    @property
+    def payment_processor(self):
+        return Liqpay(self.request.site)
+
+    # Disable atomicity for the view. Otherwise, we'd be unable to commit to the database
+    # until the request had concluded; Django will refuse to commit when an atomic() block
+    # is active, since that would break atomicity. Without an order present in the database
+    # at the time fulfillment is attempted, asynchronous order fulfillment tasks will fail.
+    @method_decorator(transaction.non_atomic_requests)
+    @method_decorator(csrf_exempt)
+    def dispatch(self, request, *args, **kwargs):
+        return super(LiqpayPaymentCallbackView, self).dispatch(request, *args, **kwargs)
+
+    def _get_basket(self, basket_id):
+        if not basket_id:
+            return None
+
+        try:
+            basket_id = int(basket_id)
+            basket = Basket.objects.get(id=basket_id)
+            basket.strategy = strategy.Default()
+            Applicator().apply(basket, basket.owner, self.request)
+            return basket
+        except (ValueError, ObjectDoesNotExist):
+            return None
+
+    def post(self, request):
+        """Handle an incoming user returned to us by LiqPay after approving payment."""
+
+        data = request.POST.get('data')
+        decode_data = json.loads(b64decode(data).decode('utf-8'))
+        basket = None
+        transaction_id = None
+
+        try:
+            transaction_id = decode_data.get('transaction_id')
+            order_number = decode_data.get('order_id')
+            basket_id = OrderNumberGenerator().basket_id(order_number)
+
+            logger.info(
+                'Received LiqPay merchant notification for transaction [%s], associated with basket [%d].',
+                transaction_id,
+                basket_id
+            )
+
+            basket = self._get_basket(basket_id)
+
+            if not basket:
+                logger.error('Received payment for non-existent basket [%s].', basket_id)
+                raise InvalidBasketError
```

Comment:
For the future, try to avoid tons of code lines in the `try` section. `Try` should always in good shape and small as much as possible.

Reply:
rewrote